[[ Bug 20504 ]] Fix __MCStringCheck function

runrevmark opened the pull request:
This function corrects the __MCStringCheck function when the
input string contains a decomposed char sequence which can
be mapped to a composed char in the native encoding.

runrevmark committed:
[[ Bug 20504 ]] Fix __MCStringCheck function

This function corrects the __MCStringCheck function when the
input string contains a decomposed char sequence which can
be mapped to a composed char in the native encoding.

Additionally, it corrects the ISO8850-1 'native pair' mapping
table ensuring that decomposable native chars in the encoding
map back to their composed counterparts.

runrevmark commented:
Okay so for some reason the test passed locally, but is failing on travis.

@livecodeali I'm concerned this fix isn't actually quite right. You asked yday about whether the 'CanBeNative' flag *should* be set in the decomposed case (where a char has a composed mapping in native) as the conversion to native is lossy. Do we still have any situations were a string will flip between unicode and native? If so, then it probably isn't quite right, so perhaps we need a separate flag (CanCompareToNative or similar). Another way to fix it is to take out the comparison culling checks. Do you have any thoughts?

livecodeali commented:
@runrevmark Once strings are unicode they will stay that way until coerced somehow, I'm pretty sure. But I don't think that matters, because I think 'CanBeNative' actually _is_ 'CanCompareToNative' (in that I believe that's the only way it is used)

runrevmark commented:
Okay - phew - I will investigate to see why this PR is failing on Travis then :)

livecodeian commented:
@runrevmark there's a failing test on Travis:
../../libfoundation/test/test_string.cpp:191: Failure
Value of: MCStringIsEqualTo(*t_decomposed, *t_composed, kMCStringOptionCompareCaseless)
  Actual: false
Expected: true
[  FAILED  ] string.normalize_compare (0 ms)

runrevmark commented:
Yes - it's somewhat odd - the same test passes on Mac. I've not managed to determine what the issue is yet :(

runrevmark commented:
Okay so the PR now passes on Travis on Linux... However, the current string flags are assuming that CanBeNative strings are Trivial - which isn't true. For example e,acute CanBeNative but is not Trivial. I think I need some help from @livecodeali to ensure we make the correct tweaks.

livecodeali commented:
@runrevmark Ah so we are conflating two different concepts of 'can be native' at the moment ☹️

runrevmark committed:
[[ Bug 20504 ]] Don't assume Native => Trivial

This patch further changes the __MCStringCheck function and also
adjusts the __MCStringSetFlags function.

Previously, there was an assumption that Native => Trivial, which
is not true in the case of there being a native composable pair
(e.g. e,acute). The __MCStringCheck function now scans the entire
string and collects whether the string is Basic, Trivial, and/or
Native. The __MCStringSetFlags function no longer assumes that
Native => Trivial.

livecodeali commented:
@livecode-vulcan review ok 9b54bb2

livecode-vulcan commented:
💙 review by @livecodeali ok 9b54bb209b305e30e1ea7cd44770ccc0b7dd05c6

livecode-vulcan commented:
😞 test failure 9b54bb209b305e30e1ea7cd44770ccc0b7dd05c6

* try-community-armv6-android-api8: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api8/builds/929)
* try-community-armv6-android-api9: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api9/builds/945)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/919)
* try-community-universal-ios-iphoneos11.0: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos11.0/builds/94)
* try-community-universal-ios-iphonesimulator11.0: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator11.0/builds/95)
* try-community-universal-mac-macosx10.6: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.6/builds/923)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/945)
* try-community-x86-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian7/builds/929)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/945)
* try-community-x86_64-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian7/builds/928)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/938)
* try-community-x86-win32: [failure](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/960)
* try-community-x86_64-win32: [pending](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/924)

livecodeali commented:
@livecode-vulcan retry 9b54bb2

livecode-vulcan commented:
😞 test failure 9b54bb209b305e30e1ea7cd44770ccc0b7dd05c6

* try-community-armv6-android-api8: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api8/builds/930)
* try-community-armv6-android-api9: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api9/builds/946)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/920)
* try-community-universal-ios-iphoneos11.0: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos11.0/builds/95)
* try-community-universal-ios-iphonesimulator11.0: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator11.0/builds/96)
* try-community-universal-mac-macosx10.6: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.6/builds/924)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/946)
* try-community-x86-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian7/builds/930)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/946)
* try-community-x86_64-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian7/builds/929)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/939)
* try-community-x86-win32: [failure](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/961)
* try-community-x86_64-win32: [success](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/925)

runrevmark committed:
[[ Bug 20504 ]] Fix wide string prefix on Win32

This patch ensures L is used on Win32 to indicate a UTF-16 codeunit
string.

runrevmark committed:
[[ Bug 20504 ]] Remove C++11isms from libfoundation string tests

This patch removes the C++11isms from the libfoundation string
tests which weren't compiling as this bug is fixed against
develop-8.1 which doesn't have C++11 support.

runrevmark committed:
[[ GoogleTests ]] Turn off C++11 support

This patch ensures that googletest doesn't compile with C++11
features which (at least when compiling develop-8.1 on newer
versions of Xcode) seem to be enabled even though the library
support isn't there.

This change can be reverted after merging up to develop - where
we should be able to enable such support.

livecodeali commented:
@livecode-vulcan review ok 583b673

livecode-vulcan commented:
💙 review by @livecodeali ok 583b67360eb680d70d78657b4c62d252c1e7dcc7

livecode-vulcan commented:
😎 test success 583b67360eb680d70d78657b4c62d252c1e7dcc7

* try-community-armv6-android-api8: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api8/builds/939)
* try-community-armv6-android-api9: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api9/builds/955)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/929)
* try-community-universal-ios-iphoneos11.1: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos11.1/builds/2)
* try-community-universal-ios-iphonesimulator11.1: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator11.1/builds/2)
* try-community-universal-mac-macosx10.6: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.6/builds/933)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/956)
* try-community-x86-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian7/builds/939)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/955)
* try-community-x86_64-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian7/builds/938)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/948)
* try-community-x86-win32: [success](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/970)
* try-community-x86_64-win32: [success](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/934)
